### PR TITLE
:sparkles: Feature:add more option and update version to 4.0.2 for reveal.js

### LINKED ISSUE #300: Update reveal.js
## Feature Request
Hi there. Hexo-theme-melody use reveal.js 3.6.0 which is out-dated. I suggest that it should be updated to 4.x.x.
## Why
Some newer plugins doesn't support v3 so you have to update this to v4
## Are you willing to work on yourself?
Yes, I'll try to update it to v4.0.2 but I don't have much spare time so this could take a long time.
I just have question: Are all the *.pug files used to configure reveal.js are come from:

* layout/includes/slide/
* source/js/thrid-party/reveal

### PR DESCRIPTION
**Pull request by LocTran016**
I have tested this, it look good.

## What to dos?

* Add all reveal.js configuarion

## Whats are missing

* Update the plugin (I don't know how to, do sorry)

It would be nice if you could take a look at this
Closed #300

## Problems

* Some configuration have some issues with them [Fixed with the lastest commit]
image
Problem with default value, I guess the way to fix this is use `false` instead of `defaultSlide.showNotes` and some others as well

### CODE REVIEW

File: layout/includes/slide/script.pug [hunk 1]
```diff
@@ -25,7 +28,11 @@ script.
     parallaxBackgroundVertical: '!{parallaxBackgroundVertical}',
     autoSlide: !{autoSlide},
     loop: !{loop},
-    autoSlideStoppable: !{autoSlideStoppable},
+    controlsLayout: '!{controlsLayout}', // Determines where controls appear, "edges" or "bottom-right"
+    controlsBackArrows: '!{controlsBackArrows}',
+    //- progress: !{progress},
```
Comment (Molunerfinn) [inline layout/includes/slide/script.pug:33]: why comment these lines

**Review by Molunerfinn — CHANGES_REQUESTED** [review]

File: layout/includes/slide/script.pug [hunk 2]
```diff
@@ -12,7 +11,11 @@ script(src=url_for("https://cdn.jsdelivr.net/npm/reveal.js@3.6.0/js/reveal.min.j
 - var parallaxBackgroundVertical = slide.parallaxBackgroundVertical || defaultSlide.parallaxBackgroundVertical
 - var autoSlide = slide.autoSlide || defaultSlide.autoSlide
 - var loop = slide.loop || defaultSlide.loop
-- var autoSlideStoppable = slide.autoSlideStoppable || defaultSlide.autoSlideStoppable
+- var controlsLayout = slide.controlsLayout || defaultSlide.controlsLayout
+- var controlsBackArrows = slide.controlsBackArrows || defaultSlide.controlsBackArrows
+- var progress = slide.progress || defaultSlide.progress
+- var showNotes = slide.showNotes || defaultSlide.showNotes
```
Comment (Molunerfinn) [inline layout/includes/slide/script.pug:17]: defaultSlide is in `_config.yml`: https://github.com/Molunerfinn/hexo-theme-melody/blob/dev/_config.yml#L162-L174

But some people update the theme without update the `melody.yml` or `_config.yml` would get wrong.

So you need to add some fallback default value. Such as:

```
var controlsLayout = slide.controlsLayout || defaultSlide.controlsLayout || 'xxxx'
var showNotes = slide.showNotes || defaultSlide.showNotes || false

// etc
```

**Review by Molunerfinn — CHANGES_REQUESTED** [review]

File: layout/includes/slide/script.pug (hunk 1 above)
Comment (Molunerfinn) [reply layout/includes/slide/script.pug:33]: L33 - L35

File: layout/includes/slide/script.pug (hunk 2 above)
Comment (Molunerfinn) [reply layout/includes/slide/script.pug:17]: And you need to add default value to the hexo-theme-melody's _config.yml just like above

File: layout/includes/slide/script.pug (hunk 1 above)
Comment (LocTran016) [reply layout/includes/slide/script.pug]: Fixed below, all of them are uncommented

File: layout/includes/slide/script.pug (hunk 2 above)
Comment (LocTran016) [reply layout/includes/slide/script.pug]: Thanks, I have added them all below

File: layout/includes/slide/layout.pug [hunk 3]
```diff
@@ -24,12 +24,14 @@ html(lang=config.language)
...
+      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.0.2/dist/reveal.css")
+      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.0.2/dist/reset.css")
       - var slide = page.slide || theme.slide
       - var defaultSlide = theme.slide
       - var slideTheme = slide.theme || defaultSlide.theme
-      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@3.6.0/css/theme/" + slideTheme + ".css")
-      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@3.6.0/lib/css/zenburn.min.css")
+      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.0.2/dist/theme/" + slideTheme + ".css")
+      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.0.2/plugin/highlight/monokai.css")
+      //- link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.0.2/plugin/highlight/zenburn.css")
```
Comment (Molunerfinn) [inline layout/includes/slide/layout.pug:34]: why not use a theme config to control this?

**Review by Molunerfinn — CHANGES_REQUESTED** [review]

File: layout/includes/slide/layout.pug (hunk 3 above)
Comment (LocTran016) [reply layout/includes/slide/layout.pug]: May be you have a look then?

File: layout/includes/slide/layout.pug [hunk 4]
```diff
@@ -24,12 +24,15 @@ html(lang=config.language)
...
     if !page.iframe
-      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@3.6.0/css/reveal.min.css")
+      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.0.2/dist/reveal.css")
+      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.0.2/dist/reset.css")
       - var slide = page.slide || theme.slide
       - var defaultSlide = theme.slide
       - var slideTheme = slide.theme || defaultSlide.theme
-      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@3.6.0/css/theme/" + slideTheme + ".css")
-      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@3.6.0/lib/css/zenburn.min.css")
+      - var slideCodeTheme = slide.codetheme || defaultSlide.codetheme || monokai
```
Comment (Molunerfinn) [inline layout/includes/slide/layout.pug:32]: `'monokai'` not `monokai`

**Review by Molunerfinn — CHANGES_REQUESTED** [review]

File: _config.yml [hunk 5]
```diff
@@ -164,6 +164,7 @@ slide:
   separator_vertical: --
   charset: utf-8
   theme: black
+  codetheme: monokai
```
Comment (Molunerfinn) [inline _config.yml:167]: Camel-Case would be better: `codeTheme`

**Review by Molunerfinn — CHANGES_REQUESTED** [review]

File: _config.yml (hunk 5 above)
Comment (LocTran016) [reply _config.yml]: Fixed below

File: layout/includes/slide/layout.pug (hunk 4 above)
Comment (Molunerfinn) [reply layout/includes/slide/layout.pug]: here~

File: layout/includes/slide/layout.pug [hunk 6]
```diff
@@ -24,12 +24,15 @@ html(lang=config.language)
...
     if !page.iframe
-      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@3.6.0/css/reveal.min.css")
+      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.0.2/dist/reveal.css")
+      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.0.2/dist/reset.css")
       - var slide = page.slide || theme.slide
       - var defaultSlide = theme.slide
       - var slideTheme = slide.theme || defaultSlide.theme
-      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@3.6.0/css/theme/" + slideTheme + ".css")
-      link(rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@3.6.0/lib/css/zenburn.min.css")
+      - var slideCodeTheme = slide.codeTheme || defaultSlide.codeTheme || monokai
```
Comment (Molunerfinn) [inline layout/includes/slide/layout.pug:32]: here, not `monokai` but `'monokai'`, `'monokai'` is a string.

**Review by Molunerfinn — CHANGES_REQUESTED** [review]

File: layout/includes/slide/layout.pug (hunk 6 above)
Comment (LocTran016) [reply layout/includes/slide/layout.pug]: Fixed below. Sorry, I forgot

**Comment by LocTran016** [comment]
Would you mind bumping version (publish the package here: https://www.npmjs.com/package/hexo-theme-melody). Thanks

File: layout/includes/slide/layout.pug (hunk 4 above)
Comment (LocTran016) [reply layout/includes/slide/layout.pug]: Fixed below

File: _config.yml [hunk 7]
```diff
@@ -172,6 +173,14 @@ slide:
   parallaxBackgroundSize: ""
   parallaxBackgroundHorizontal: null
   parallaxBackgroundVertical: null
+  loop: false
+  autoSlide: false
+  controlsBackArrows: "faded"
```
Comment (Molunerfinn) [inline _config.yml:178]: use a single quote, the same as bellow

**Review by Molunerfinn — CHANGES_REQUESTED** [review]

File: _config.yml (hunk 7 above)
Comment (LocTran016) [reply _config.yml]: Change them all

**Comment by Molunerfinn** [comment]
Thanks for your contribution. While using `npm run cz` to commit next time. You should run `npm install` in the first time.

**Comment by Molunerfinn** [comment]
hexo-theme-melody uses the [PicGo/bump-version](https://github.com/PicGo/bump-version) to commit

**Comment by LocTran016** [comment]
> Thanks for your contribution. While using `npm run cz` to commit next time. You should run `npm install` in the first time.
I haven't known this before, I just use `git commit -m` and `git push`. Thanks for letting me know

**Pull request merged**